### ISSUE TITLE
Document the limitations

### ISSUE THREAD

**Issue by jtpio** [jtpio/jupyterlite#199]
Maybe we could add a quick section to the docs to document (current) limitations with JupyterLite.

Some folks might think this is exactly the same as stock JupyterLab, especially regarding the Python kernel.

Maybe there could also be a link to the Pyodide limitations if any, or https://pyodide.org/en/latest/usage/wasm-constraints.html.

cc @ibdafna who might be interested in this

**Comment by bollwyvl** [comment]
Maybe in a Tour #31?

**Comment by psychemedia** [comment]
Would a list of known unsupported packages that are widely used dependencies be useful? Eg I don't think cairo is part of the wasm build so `pycairo` etc won't be useable.

**Comment by jtpio** [comment]
Sure. Those could go to the kernel specific documentation pages.

**Comment by qci-amos** [comment]
I discovered jupyter lite yesterday and it looks great!

I've got some of my own third-party python packages (in private repos) that I'd like to include. I can't find any documentation about this (e.g., it doesn't _seem_ like this is an "extension", but maybe I'm misreading what I did find). Is there a way to do this or is it a limitation?

This is the closest discussion I found: https://pyodide.org/en/stable/development/new-packages.html

**Comment by jtpio** [comment]
@qci-amos if your Python pacakges are pure Python wheels, you can probably host them on the same server as where JupyterLite is deployed. And then use the full url with `micropip`, for example `await micropip.install('https://<url-to-wheel>')`.

If the Python package requires some extensions to also be installed on the frontend (for example if it's a Jupyter Widget), then you might need to follow this to make sure they are included in a `jupyter lite build`: https://jupyterlite.readthedocs.io/en/latest/configuring.html#adding-extensions

**Comment by qci-amos** [comment]
I'm just dabbling for now, don't have a server, all I did was the local `jupyter lite serve`... but I think what you mean is make a wheel, make it available via a url somewhere (presumably somewhere within the jupyter lite build is ok?), launch the server, and then execute that line from within a notebook?

**Comment by jtpio** [comment]
> presumably somewhere within the jupyter lite build is ok?

For now there is no real tooling around wheels (discussed in #225), but a URL might work. Users will have to write the following in their notebooks:

```python
import micropip
await micropip.install('url-to-the-python-package-as-wheel')
```

**Comment by psychemedia** [comment]
I've been wondering whether a a quick fix might to have a path `/wheels` as well as `/lab` and `/retro/tree` that universal wheel packages could be placed in (or even built into as part of `jupyter lite build` that could then be loaded via the `micropip.install(URL)` route.

But then, if you do this, I wonder what would prevent you actuall installling the package into the built environment directly? (I don't really have a very good model/understanding at all of what is going on under the hood!)

**Comment by ibdafna** [comment]
Theoretically you could have a list of wheels to `micropip` when the distribution loads, but it only covers the problem with bandaid - it doesn't fix the underlying issue. The reality is that all packages, ideally, can be built with pyodide and distributed via the CDN.

**Comment by datakurre** [comment]
@bollwyvl A question, I'd like to ask in context of this issue: should it be already possible to refer to other files visible in JupyterLite file browser? Like import a custom Python module? Or does every kernel instance run in its own empty sandbox for reason?

**Comment by bollwyvl** [comment]
Yep the general "get stuff from contents provider x into kernel y" upstream
issue is still open, unfortunately.

We _had_ to solve the build-time python package use case, if only so we
could confidently ship the kernel itself, but a general solution is still
elusive for arbitrary contents: just copying _everything_ will fail
catastrophically in a fair number of cases, and the bidirectional case is
weird, too.

There was a compelling example of an emscripten FS driver shared by the
starboard folks, but I haven't had the opportunity to figure out how to
integrate it into lite. The FS stuff is tricky, but could theoretically
allow for kernel startup customization by normal IPython means, which has
also been frequently requested.
